### ISSUE TITLE
relaxation variable r in RES Diffuser

### ISSUE THREAD

**Issue by L1U-C** [Weixy21/SafeDiffuser#19]
In the section 3.2 RELAXED-SAFE (RES) DIFFUSER, a relaxation variable r is introduced to address the local trap issue. I believe the role of r is to relax the safety constraints of CBF (formula 10) during the initial denoising stage, but this relaxation of safety constraints can only be achieved when the value of r is negative. 
However, the corresponding QP problem (formula 14) involves ||r||^2, which does not impose any requirement on the sign of r. If the solver obtains a positive value for r, it is equivalent to adopting a stricter form of CBF safety constraints, and the purpose of relaxing the constraints cannot be achieved. 
Could you please explain this issue.

The corresponding code for RES is as follows:

` @torch.no_grad()   #only for sampling
    def invariance_relax(self, x, xp1, t):  #  ReS-diffuser for maze2d-large-v1

        x = x.squeeze(0)
        xp1 = xp1.squeeze(0)

        nBatch = x.shape[0]
        ref = xp1 - x

        #normalize obstacle 1, x-1, y-0  x = 1/12*np.cos(theta) + 5.5/12, y = 1/9*np.sin(theta) + 5/9
        xr = 2*1/(self.norm_maxs[1] - self.norm_mins[1])
        yr = 2*1/(self.norm_maxs[0] - self.norm_mins[0])
        off_x = 2*(5.8-0.5 - self.norm_mins[1])/(self.norm_maxs[1] - self.norm_mins[1]) - 1
        off_y = 2*(5-0.5 - self.norm_mins[0])/(self.norm_maxs[0] - self.norm_mins[0]) - 1

        #CBF
        b = ((x[:,2:3] - off_y)/yr)**2 + ((x[:,3:4] - off_x)/xr)**2 - 1 - 0.01
        Lfb = 0
        Lgbu1 = 2*((x[:,2:3] - off_y)/yr)/yr
        Lgbu2 = 2*((x[:,3:4] - off_x)/xr)/xr

        self.safe1 = torch.min(b[:,0] + 0.01)

        if t >= 10:   # debug  10
            sign = 100   #relax
        else:
            sign = 0   #non-relax

        rx0 = torch.zeros_like(Lgbu1).to(b.device)
        rx1 = sign*torch.ones_like(Lgbu1).to(b.device)

        G = torch.cat([-Lgbu1, -Lgbu2, rx1, rx0], dim = 1)
        G = G.unsqueeze(1)
        k = 1
        h = Lfb + k*b

        #normalize obstacle 2,  x = 1/12*np.sqrt(np.abs(np.cos(theta)))*np.sign(np.cos(theta)) + 5.3/12, y = 1/9*np.sqrt(np.abs(np.sin(theta)))*np.sign(np.sin(theta)) + 2/9
        xr = 2*1/(self.norm_maxs[1] - self.norm_mins[1])
        yr = 2*1/(self.norm_maxs[0] - self.norm_mins[0])
        off_x = 2*(5.3-0.5 - self.norm_mins[1])/(self.norm_maxs[1] - self.norm_mins[1]) - 1
        off_y = 2*(2-0.5 - self.norm_mins[0])/(self.norm_maxs[0] - self.norm_mins[0]) - 1

        #CBF
        b = ((x[:,2:3] - off_y)/yr)**4 + ((x[:,3:4] - off_x)/xr)**4 - 1 - 0.01
        Lfb = 0
        Lgbu1 = 4*((x[:,2:3] - off_y)/yr)**3/yr
        Lgbu2 = 4*((x[:,3:4] - off_x)/xr)**3/xr

        self.safe2 = torch.min(b[:,0] + 0.01)

        G1 = torch.cat([-Lgbu1, -Lgbu2, rx0, rx1], dim = 1)
        G1 = G1.unsqueeze(1)
        k = 1
        h1 = Lfb + k*b

        G = torch.cat([G, G1], dim = 1)
        h = torch.cat([h, h1], dim = 1)
        
   
        q = -ref[:,2:4].to(G.device)
        q0 = torch.zeros_like(q).to(G.device)
        q = torch.cat([q, q0], dim = 1)
        Q = Variable(torch.eye(4))
        Q = Q.unsqueeze(0).expand(nBatch, 4, 4).to(G.device)
        
        e = Variable(torch.Tensor())
        out = QPFunction(verbose=-1, solver = QPSolvers.PDIPM_BATCHED)(Q, q, G, h, e, e)

        rt = xp1.clone()      
        rt[:,2:4] = x[:,2:4] + out[:,0:2]
        rt = rt.unsqueeze(0)
        return rt`

**Comment by Weixy21** [comment]
The ReS-Diffuser only relax the constraints for the first M time steps, and the remaining N-M time steps still take the constraints as hard ones, i.e., without the slack variable \delta

Below is the code, when sign = 0, then the constraint is a hard one.

if t >= 10:   # debug  10
        sign = 100   #relax
    else:
        sign = 0   #non-relax

**Comment by L1U-C** [comment]
> The ReS-Diffuser only relax the constraints for the first M time steps, and the remaining N-M time steps still take the constraints as hard ones, i.e., without the slack variable \delta
> 
> Below is the code, when sign = 0, then the constraint is a hard one.
> 
> if t >= 10: # debug 10 sign = 100 #relax else: sign = 0 #non-relax

Yes, I am well aware of that. What truly puzzles me is actually the sign of the r. In my understanding, r should be a negative value. Is that correct?

**Comment by Weixy21** [comment]
Yes, the constraint is relaxed when r is negative. When r is positive, then the constraint becomes more stringent. Although the cost takes r^2, r will not become positive in practice.  This is some default rule in the CBF-based QP. Of course, you can add an additional constraint that r\leq 0, but this won't affect the solution.

**Comment by L1U-C** [comment]
> Yes, the constraint is relaxed when r is negative. When r is positive, then the constraint becomes more stringent. Although the cost takes r^2, r will not become positive in practice. This is some default rule in the CBF-based QP. Of course, you can add an additional constraint that r\leq 0, but this won't affect the solution.

Thank you for your clarification. 
I am not familiar with the knowledge of CBF-based QP. Is the conclusion that r is not a positive number derived from a certain theory or other analytical methods? Sorry for my confusion about this question

**Comment by Weixy21** [comment]
It is not from any theory, but from empirical results. This may be done by some analysis.

**Comment by L1U-C** [comment]
> It is not from any theory, but from empirical results. This may be done by some analysis.

Well, this problem is a little confusing for me. Can you provide a simple guide, or I will continue to study this problem for a while. Anyway, thank you for your help.

**Comment by Weixy21** [comment]
You may find the closed-form solution of the QP when there is a slack variable, and analysis it based on the closed-form solution. The closed-form solution is given in the appendix (without the slack variable)
